roles/download/tasks/download_container.yml

```diff
@@ -1,40 +1,125 @@
   when:
-    - download_run_once
-    - download.enabled
-    - download.container
-    - any_pull_required | default(download_always_pull)
-  delegate_to: "{{ download_delegate }}"
-  delegate_facts: yes
-  run_once: yes
+    - download_force_cache
+    - image_is_cached
+    - not download_localhost
+    - ansible_os_family not in ["CoreOS", "Container Linux by CoreOS"]
+  tags:
+    - upload
+    - upgrade
+
+# NOTE: If download_delegate==localhost, then the account running ansible must be in the group 'docker' on the ansible host
+- name: download_container | Load image into docker
+  shell: "{{ docker_bin_dir }}/docker load < {{ image_path_cached if download_localhost else image_path_final }}"
+  delegate_to: "{{ download_delegate if download_run_once or inventory_hostname }}"
+  run_once: "{{ download_run_once }}"
+  register: container_load_status
+  failed_when: container_load_status | failed
+  become: "{{ not download_localhost }}"
+  when:
+    - image_is_cached
+    - download_force_cache
+    - ansible_os_family not in ["CoreOS", "Container Linux by CoreOS"]
+  tags:
+    - upload
+    - upgrade
+
+- name: download_container | Prepare container download
+  import_tasks: check_pull_required.yml
+  run_once: "{{ download_run_once }}"
+  when:
+    - not download_always_pull
 
-- name: container_download | Download containers if pull is required or told to always pull (all nodes)
-  command: "{{ docker_bin_dir }}/docker pull {{ pull_args }}"
+- debug:
+    msg: "XXX Pull required is: {{ pull_required }}"
+
+# NOTE: Pre-loading docker images will not prevent 'docker pull' from re-downloading the layers in that image
+# if a pull is forced. This is a known issue with docker. See https://github.com/moby/moby/issues/23684
+- name: download_container | Download image if required
+  command: "{{ docker_bin_dir }}/docker pull {{ image_reponame }}"
+  delegate_to: "{{ download_delegate if download_run_once or inventory_hostname }}"
+  delegate_facts: yes
+  run_once: "{{ download_run_once }}"
   register: pull_task_result
   until: pull_task_result is succeeded
-  retries: 4
   delay: "{{ retry_stagger | random + 3 }}"
-  changed_when: not 'up to date' in pull_task_result.stdout
+  retries: 4
+  become: "{{ not download_localhost }}"
+  when:
+    - pull_required | default(download_always_pull)
+  tags:
+    - upload
+    - upgrade
+
+# NOTE: image_changed is only valid if a pull is was needed or forced.
+- name: download_container | Check if image changed
+  set_fact:
+    image_changed: "{{ true if pull_task_result.stdout is defined and not 'up to date' in pull_task_result.stdout else false }}"
+  run_once: true
+  when:
+    - download_force_cache or download_run_once
+  tags:
+    - facts
+
+- name: download_container | Save and compress image
+  shell: "{{ docker_bin_dir }}/docker save {{ image_reponame }} | gzip -{{ download_compress }} > {{ image_path_cached if download_localhost else image_path_final }}"
```

Comment:
wouldn't https://docs.ansible.com/ansible/2.7/modules/docker_image_module.html cut it ?

Reply:
It might, partially, though the module does not seem to support compression. That would require an extra command to gzip and ungzip the images. Compression is quite desirable as it reduces the current size of the cache from 1.6G to 540M. And even if the module would fit the requirements, it should be done is a seperate PR, as noted in  a previous reply.

Reply:
fair point